src/Grpc.Net.Client/Internal/Retry/RetryCall.cs

```diff
@@ -0,0 +1,320 @@
+                        retryPushbackMS < 0)
+                    {
+                        Channel.RetryThrottling?.CallFailure();
+                    }
+
+                    var result = EvaluateRetry(status, retryPushbackMS);
+                    Log.RetryEvaluated(Logger, status.StatusCode, AttemptCount, result == null);
+
+                    if (result == null)
+                    {
+                        TimeSpan delayDuration;
+                        if (retryPushbackMS != null)
+                        {
+                            delayDuration = TimeSpan.FromMilliseconds(retryPushbackMS.GetValueOrDefault());
+                            _nextRetryDelayMilliseconds = retryPushbackMS.GetValueOrDefault();
+                        }
+                        else
+                        {
+                            delayDuration = TimeSpan.FromMilliseconds(Channel.GetRandomNumber(0, Convert.ToInt32(_nextRetryDelayMilliseconds)));
+                        }
+
+                        Log.StartingRetryDelay(Logger, delayDuration);
+                        await Task.Delay(delayDuration, CancellationTokenSource.Token).ConfigureAwait(false);
+
+                        _nextRetryDelayMilliseconds = CalculateNextRetryDelay();
+
+                        // Check if dispose was called on call.
+                        CancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                        // Clean up the failed call.
+                        currentCall.Dispose();
+                    }
+                    else
+                    {
+                        // Handle the situation where the call failed with a non-deadline status, but retry
+                        // didn't happen because of deadline exceeded.
+                        IGrpcCall<TRequest, TResponse> resolvedCall = (IsDeadlineExceeded() && !(currentCall.CallTask.IsCompletedSuccessfully() && currentCall.CallTask.Result.StatusCode == StatusCode.DeadlineExceeded))
+                            ? CreateStatusCall(GrpcProtocolConstants.DeadlineExceededStatus)
+                            : currentCall;
+
+                        // Can't retry.
+                        // Signal public API exceptions that they should finish throwing and then exit the retry loop.
+                        CommitCall(resolvedCall, result.GetValueOrDefault());
+                        return;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                HandleUnexpectedError(ex);
+            }
+            finally
+            {
+                Log.StoppingRetryWorker(Logger);
+            }
+        }
+
+        protected override void OnCommitCall(IGrpcCall<TRequest, TResponse> call)
+        {
+            _activeCall = null;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            lock (Lock)
+            {
+                base.Dispose(disposing);
```

Comment:
Hmm? Doesn't `base.Dispose` also lock when disposing is true?

Reply:
Yes. It is the same lock so it doesn't matter.